src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+        computeLocalTime(commandPath.mEndpointId);
+        return true;
+    }
+
+    computeLocalTime(commandPath.mEndpointId);
+
+    commandObj->AddStatus(commandPath, Status::Success);
+
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetDSTOffsetCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetDSTOffset::DecodableType & commandData)
+{
+    Status status  = Status::Success;
+    auto dstOffset = commandData.DSTOffset;
+    size_t items;
+    uint8_t maxSize = 0;
+    dstOffset.ComputeSize(&items);
+    DSTOffsetListMaxSize::Get(commandPath.mEndpointId, &maxSize);
+
+    if (items > maxSize && items > CHIP_CONFIG_DST_OFFSET_LIST_MAX_SIZE)
+    {
+        commandObj->AddStatus(commandPath, Status::ResourceExhausted);
+        return true;
+    }
+
+    // sorted by ValidStarting time
+    // ValidStartingTime shall not be smaller than ValidUntil time of previous entry
+    // only 1 validuntil null value and shall be last in the list
+    // remove entries which are no longer active
+    // if offset == 0 && ValidUntil == null then no DST is used
+    if (CHIP_NO_ERROR != TimeSynchronizationServer::Instance().SetDSTOffset(dstOffset))
+    {
+        commandObj->AddStatus(commandPath, Status::ConstraintError);
+        return true;
+    }
+    // if DST state changes, generate DSTStatus event
+    sendDSTStatus(commandPath.mEndpointId, true);
+    // if list is empty, generate DSTTableEmpty event
+    sendDSTTableEmpty(commandPath.mEndpointId);
+
+    commandObj->AddStatus(commandPath, status);
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetDefaultNTPCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetDefaultNTP::DecodableType & commandData)
+{
+    Status status = Status::Success;
+    auto dntpChar = commandData.defaultNTP;
+    DataModel::Nullable<chip::MutableByteSpan> dntpByte;
+
+    if (!dntpChar.IsNull())
+    {
+        if (!GetDelegate()->isNTPAddressValid(dntpChar.Value()))
+        {
+            commandObj->AddStatus(commandPath, Status::InvalidCommand);
+            return true;
+        }
+        if (GetDelegate()->isNTPAddressDomain(dntpChar.Value()))
+        {
+            bool dnsResolve;
+            SupportsDNSResolve::Get(commandPath.mEndpointId, &dnsResolve);
+            if (!dnsResolve)
+            {
+                commandObj->AddStatus(commandPath, Status::InvalidCommand);
+                return true;
+            }
+        }
+
+        uint8_t buffer[DefaultNTP::TypeInfo::MaxLength()];
+        chip::MutableByteSpan dntp(buffer);
+        size_t len = (dntpChar.Value().size() < dntp.size()) ? dntpChar.Value().size() : dntp.size();
+        memcpy(buffer, dntpChar.Value().data(), len);
```

Comment:
Why are we just dropping the end of the string if it's too long, instead of replying with a failure?  If it's too long, client violated the constraint, no?